cloud-openstack-server/src/main/java/jetbrains/buildServer/clouds/openstack/OpenstackCloudInstance.java

```diff
@@ -63,45 +63,56 @@ public OpenstackCloudInstance(@NotNull final OpenstackCloudImage image, @NotNull
-                case BUILD:
-                case REBUILD:
-                    setStatus(InstanceStatus.STARTING);
-                    break;
-                case ACTIVE:
-                    // When OpenStack instance is stopping, the status is always 'ACTIVE' => check if termination started
-                    if (InstanceStatus.SCHEDULED_TO_STOP.equals(getStatus()) || InstanceStatus.STOPPING.equals(getStatus())) {
-                        setStatus(InstanceStatus.STOPPING);
-                    } else {
-                        setStatus(InstanceStatus.RUNNING);
+            try {
+                Server server = cloudImage.getNovaServerApi().get(serverCreated.getId());
+                if (server != null) {
+                    Server.Status currentStatus = server.getStatus();
+                    LOG.debug(String.format("Getting instance status from openstack for '%s', result is '%s' (previous internal status setted was '%s')",
+                            getName(), currentStatus, getStatus()));
+                    switch (currentStatus) {
+                        case BUILD:
+                        case REBUILD:
+                            setStatus(InstanceStatus.STARTING);
+                            break;
+                        case ACTIVE:
+                            // When OpenStack instance is stopping, the status is always 'ACTIVE' => check if termination started
+                            if (InstanceStatus.SCHEDULED_TO_STOP.equals(getStatus()) || InstanceStatus.STOPPING.equals(getStatus())) {
+                                setStatus(InstanceStatus.STOPPING);
+                            } else {
+                                setStatus(InstanceStatus.RUNNING);
+                            }
+                            break;
+                        case ERROR:
+                            setStatus(InstanceStatus.ERROR);
+                            terminate();
+                            break;
+                        case SHUTOFF:
+                            terminate();
+                            break;
+                        case DELETED:
+                        case SUSPENDED:
+                        case PAUSED:
+                        case SOFT_DELETED:
+                        case UNKNOWN:
+                        case UNRECOGNIZED:
+                        case SHELVED:
+                        case SHELVED_OFFLOADED:
+                        default:
+                            setStatus(InstanceStatus.STOPPED);
+                            break;
                     }
-                    break;
-                case ERROR:
-                    terminate();
+                } else {
+                    setStatus(InstanceStatus.STOPPED);
+                }
+            } catch (Exception e) {
+                LOG.error("Got exception while calculating instance status, will terminate", e);
+                try {
                     setStatus(InstanceStatus.ERROR);
```

Comment:
This status and the following `try/catch` are probably useless.
The [terminate](https://github.com/yandex-qatools/teamcity-openstack-plugin/blob/c56f7989c640b950fd2de5e62663941f89c9cc1e/cloud-openstack-server/src/main/java/jetbrains/buildServer/clouds/openstack/OpenstackCloudInstance.java#L189) method manages the status and any error if throwed.